src/ir/lowering/mod.rs

```diff
@@ -968,15 +985,27 @@ impl<'k> LowerGoal<Env<'k>> for Goal {
                 // `if (FromEnv(T: Trait)) { ... /* this part is untouched */ ... }`.
                 let where_clauses: Result<Vec<_>> =
                     wc.into_iter()
-                      .map(|wc| Ok(wc.lower_clause(env)?.into_from_env_clause()))
+                      .flat_map(|wc| wc.lower_clause(env).apply_result())
+                      .map(|result| result.map(|wc| wc.into_from_env_clause()))
                       .collect();
                 Ok(Box::new(ir::Goal::Implies(where_clauses?, g.lower(env)?)))
             }
             Goal::And(g1, g2) => {
                 Ok(Box::new(ir::Goal::And(g1.lower(env)?, g2.lower(env)?)))
             }
             Goal::Not(g) => Ok(Box::new(ir::Goal::Not(g.lower(env)?))),
-            Goal::Leaf(wc) => Ok(Box::new(ir::Goal::Leaf(wc.lower(env)?))),
+            Goal::Leaf(wc) => {
+                // A where clause can lower to multiple leaf goals; wrap these in Goal::And.
```

Comment:
Not certain about this yet

Reply:
Seems good to me! Also you can use `itertools::Itertools` and write something like:
```rust
leaves.fold1(|goal, leaf| ir::Goal::And(Box::new(goal), Box::new(leaf)))
      .expect("at least one goal");
```
(see for example [here](https://github.com/rust-lang-nursery/chalk/blob/master/src/rules/wf/mod.rs#L265))

Reply:
Bah, itertools should have called this method `reduce`. =)